### ISSUE TITLE
ControllerTest::testThemeUrl failure

### ISSUE THREAD

**Issue by alen12345** [octobercms/october#2939]
##### Expected behavior
Test should retrieve the correct theme path.

##### Actual behavior
The theme path is tested against a localhost path.

##### Reproduce steps
Run *phpunit*

##### October build
419

##### Generated output
`There was 1 failure:

1) ControllerTest::testThemeUrl
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'http://localhost/themes/test'
+'https://example.com.it/themes/test'

/var/www/html/meccMilano/tests/unit/cms/classes/ControllerTest.php:23`

**Comment by LukeTowers** [comment]
Did you set up your config/app.php correctly?

**Comment by alen12345** [comment]
This is my configuration in the congig/app.php:
`'url' => env('APP_URL', 'https://example1.it'),`

where example1.it is the production server. This is however overriden by the APP_URL string in the .env file, where for example in development I have:
`APP_URL=http://localhost:8080/meccMilanoDevelopment`

The error reported is as follows:
`There was 1 failure:

1) ControllerTest::testThemeUrl
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'http://localhost/themes/test'
+'http://localhost:8080/meccMilanoDevelopment/themes/test'`

Please tell me if I am missing something in the configuration.

Thank you for your help.

**Comment by harmenjanssen** [comment]
I can confirm this issue.  
I just did a fresh install (using `composer create-project`) and phpunit fails on this test.

Both `config/app.php` and `.env` contain a custom URL:

```php
'url' => env('APP_URL', 'http://localhost.stedelijk.nl')
```

I feel like I'm missing something – the test is looking at a hard-coded path, should it be configured in `setup` maybe?

**Comment by HammenWS** [comment]
@alen12345 @harmenjanssen @LukeTowers 

Confirmed!

But you can make that specific test pass by adding the following env variable to the `/phpunit.xml`

```
<php>
        <env name="APP_ENV" value="testing" />
        <env name="APP_URL" value="http://localhost" /> // add this line
        <env name="CACHE_DRIVER" value="array" />
        <env name="SESSION_DRIVER" value="array" />
</php>
```

**Comment by harmenjanssen** [comment]
Of course! Thanks @HammenWS that fixed it for me.  
It makes sense to setup your testing ENV like that, this would allow you to configure a mock database as well.

This is worthy of a pull request I'd say 🙂

**Comment by alen12345** [comment]
@HammenWS @harmenjanssen That should be definitely merged!

**Comment by LukeTowers** [comment]
Leaving it open until the issue is actually corrected with a code change merged into core.

**Comment by alen12345** [comment]
On second thoughts, and having looked at the code, I think that relying on env vars for a test or hardcoding an url is really bad practice. I will look at PR #3074 and think about a workaround.

**Comment by alen12345** [comment]
@LukeTowers can you please have a look at my proposal at PR #3074 ? I think we can fix the issue that way.

**Comment by w20k** [comment]
Will be fixed as a part of PR #4378
